### ISSUE TITLE
🔥 [🐛] Unable to initialize Mobile Ads SDK

### ISSUE THREAD

**Issue by shaun-ps-04** [invertase/react-native-google-mobile-ads#158]
## Issue

Expo SDK: v45

I have an expo app that is failing to load the Mobile Ads SDK in development (haven't tested production build). It fails when attempting to initialize, returning the following error:
> [TypeError: null is not an object (evaluating 'RNAppModule.eventsNotifyReady')]

I think I found the file its failing on [here](https://github.com/invertase/react-native-google-mobile-ads/blob/469b3be4197545c0c6f1da248539d94b864d9be9/jest.setup.ts).

`RNAppModule` does not have an attribute called `eventsNotifyReady` however the attribute below, `RNGoogleMobileAdsModule` does have `eventsNotifyReady`.

Hopefully this helps because I am unsure what to do with this information.

In case this is of any interest, to stop the web version of my app crashing I am conditionally importing the package like so:

```javascript
const mobileAdsPromise = async () => {
  if ( ["android", "ios"].includes(Platform.OS) ) {
    return await import("react-native-google-mobile-ads");
  } else null;
}
```

And this is how I am initializing the SDK:

```javascript
mobileAdsPromise()
  .then( res => {
    console.log("\nres :", res);
    console.log("\nres.MobileAds :", res.MobileAds );
    console.log("\nres.MobileAds() :", res.MobileAds() );
    res.MobileAds()
      .initialize()
      .then( adapterStatuses => {
        console.log("Adapter Statuses :", adapterStatuses);
      }).catch( err => {
        console.log("Adapter Statuses Err:", err);
      })
  }).catch( err => {
    console.log("Mobile Ads Err :", err);
  })
```

### Here are the other logs:
<details><summary>Click To Expand</summary>

```
res : Object {
  "AdEventType": Object {
    "CLICKED": "clicked",
    "CLOSED": "closed",
    "ERROR": "error",
    "LOADED": "loaded",
    "OPENED": "opened",
  },
  "AdsConsent": Object {
    "getTCModel": [Function getTCModel],
    "getTCString": [Function getTCString],
    "getUserChoices": [Function getUserChoices],
    "requestInfoUpdate": [Function requestInfoUpdate],
    "reset": [Function reset],
    "showForm": [Function showForm],
  },
  "AdsConsentDebugGeography": Object {
    "0": "DISABLED",
    "1": "EEA",
    "2": "NOT_EEA",
    "DISABLED": 0,
    "EEA": 1,
    "NOT_EEA": 2,
  },
  "AdsConsentPurposes": Object {
    "1": "STORE_AND_ACCESS_INFORMATION_ON_DEVICE",
    "10": "DEVELOP_AND_IMPROVE_PRODUCTS",
    "2": "SELECT_BASIC_ADS",
    "3": "CREATE_A_PERSONALISED_ADS_PROFILE",
    "4": "SELECT_PERSONALISED_ADS",
    "5": "CREATE_A_PERSONALISED_CONTENT_PROFILE",
    "6": "SELECT_PERSONALISED_CONTENT",
    "7": "MEASURE_AD_PERFORMANCE",
    "8": "MEASURE_CONTENT_PERFORMANCE",
    "9": "APPLY_MARKET_RESEARCH_TO_GENERATE_AUDIENCE_INSIGHTS",
    "APPLY_MARKET_RESEARCH_TO_GENERATE_AUDIENCE_INSIGHTS": 9,
    "CREATE_A_PERSONALISED_ADS_PROFILE": 3,
    "CREATE_A_PERSONALISED_CONTENT_PROFILE": 5,
    "DEVELOP_AND_IMPROVE_PRODUCTS": 10,
    "MEASURE_AD_PERFORMANCE": 7,
    "MEASURE_CONTENT_PERFORMANCE": 8,
    "SELECT_BASIC_ADS": 2,
    "SELECT_PERSONALISED_ADS": 4,
    "SELECT_PERSONALISED_CONTENT": 6,
    "STORE_AND_ACCESS_INFORMATION_ON_DEVICE": 1,
  },
  "AdsConsentSpecialFeatures": Object {
    "1": "USE_PRECISE_GEOLOCATION_DATA",
    "2": "ACTIVELY_SCAN_DEVICE_CHARACTERISTICS_FOR_IDENTIFICATION",
    "ACTIVELY_SCAN_DEVICE_CHARACTERISTICS_FOR_IDENTIFICATION": 2,
    "USE_PRECISE_GEOLOCATION_DATA": 1,
  },
  "AdsConsentStatus": Object {
    "NOT_REQUIRED": "NOT_REQUIRED",
    "OBTAINED": "OBTAINED",
    "REQUIRED": "REQUIRED",
    "UNKNOWN": "UNKNOWN",
  },
  "AppOpenAd": [Function AppOpenAd],
  "BannerAd": [Function BannerAd],
  "BannerAdSize": Object {
    "ADAPTIVE_BANNER": "ADAPTIVE_BANNER",
    "ANCHORED_ADAPTIVE_BANNER": "ANCHORED_ADAPTIVE_BANNER",
    "BANNER": "BANNER",
    "FLUID": "FLUID",
    "FULL_BANNER": "FULL_BANNER",
    "INLINE_ADAPTIVE_BANNER": "INLINE_ADAPTIVE_BANNER",
    "LARGE_BANNER": "LARGE_BANNER",
    "LEADERBOARD": "LEADERBOARD",
    "MEDIUM_RECTANGLE": "MEDIUM_RECTANGLE",
    "WIDE_SKYSCRAPER": "WIDE_SKYSCRAPER",
  },
  "GAMAdEventType": Object {
    "APP_EVENT": "app_event",
  },
  "GAMBannerAd": [Function GAMBannerAd],
  "GAMInterstitialAd": [Function GAMInterstitialAd],
  "InitializationState": Object {
    "0": "AdapterInitializationStateNotReady",
    "1": "AdapterInitializationStateReady",
    "AdapterInitializationStateNotReady": 0,
    "AdapterInitializationStateReady": 1,
  },
  "InterstitialAd": [Function InterstitialAd],
  "MaxAdContentRating": Object {
    "G": "G",
    "MA": "MA",
    "PG": "PG",
    "T": "T",
  },
  "MobileAds": [Function MobileAds],
  "RewardedAd": [Function RewardedAd],
  "RewardedAdEventType": Object {
    "EARNED_REWARD": "rewarded_earned_reward",
    "LOADED": "rewarded_loaded",
  },
  "RewardedInterstitialAd": [Function RewardedInterstitialAd],
  "SDK_VERSION": "6.3.0",
  "TestIds": Object {
    "APP_OPEN": "ca-app-pub-3940256099942544/3419835294",
    "BANNER": "ca-app-pub-3940256099942544/6300978111",
    "GAM_APP_OPEN": "/6499/example/app-open",
    "GAM_BANNER": "/6499/example/banner",
    "GAM_INTERSTITIAL": "/6499/example/interstitial",
    "GAM_NATIVE": "/6499/example/native",
    "GAM_REWARDED": "/6499/example/rewarded",
    "GAM_REWARDED_INTERSTITIAL": "/21775744923/example/rewarded_interstitial",
    "INTERSTITIAL": "ca-app-pub-3940256099942544/1033173712",
    "REWARDED": "ca-app-pub-3940256099942544/5224354917",
    "REWARDED_INTERSTITIAL": "ca-app-pub-3940256099942544/5354046379",
  },
  "default": [Function MobileAds],
  "useAppOpenAd": [Function useAppOpenAd],
  "useInterstitialAd": [Function useInterstitialAd],
  "useRewardedAd": [Function useRewardedAd],
  "useRewardedInterstitialAd": [Function useRewardedInterstitialAd],
}

res.MobileAds : [Function MobileAds]

res.MobileAds() : MobileAdsModule {
  "_app": Object {
    "name": "AppName",
  },
  "_config": Object {
    "namespace": "google_mobile_ads",
    "nativeEvents": Array [
      "google_mobile_ads_app_open_event",
      "google_mobile_ads_interstitial_event",
      "google_mobile_ads_rewarded_event",
      "google_mobile_ads_rewarded_interstitial_event",
    ],
    "nativeModuleName": Array [
      "RNGoogleMobileAdsModule",
      "RNGoogleMobileAdsAppOpenModule",
      "RNGoogleMobileAdsInterstitialModule",
      "RNGoogleMobileAdsRewardedModule",
      "RNGoogleMobileAdsRewardedInterstitialModule",
    ],
    "version": "6.3.0",
  },
  "_nativeModule": null,
}

Mobile Ads Err : [TypeError: null is not an object (evaluating 'RNAppModule.eventsNotifyReady')]
```
</details>

---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
{
  "name": "***",
  "version": "1.2.1",
  "main": "node_modules/expo/AppEntry.js",
  "scripts": {
    "start": "expo start",
    "android": "expo start --android",
    "ios": "expo start --ios",
    "web": "expo start --web",
    "eject": "expo eject",
    "build": "expo build:web"
  },
  "dependencies": {
    "@expo-google-fonts/nunito": "^0.2.2",
    "@expo/config-plugins": "^4.1.5",
    "@react-native-async-storage/async-storage": "~1.17.3",
    "@react-navigation/material-bottom-tabs": "^6.2.1",
    "@react-navigation/native": "^6.0.10",
    "@react-navigation/native-stack": "^6.6.2",
    "@react-navigation/stack": "^6.2.1",
    "axios": "^0.27.2",
    "babel-plugin-root-import": "^6.6.0",
    "dayjs": "^1.11.2",
    "expo": "^45.0.1",
    "expo-blur": "~11.1.0",
    "expo-image-manipulator": "~10.3.1",
    "expo-image-picker": "~13.1.1",
    "expo-linear-gradient": "~11.3.0",
    "expo-updates": "~0.13.2",
    "jwt-decode": "^3.1.2",
    "react": "17.0.2",
    "react-dom": "17.0.2",
    "react-native": "0.68.2",
    "react-native-awesome-alerts": "^1.5.2",
    "react-native-circular-progress": "^1.3.7",
    "react-native-gesture-handler": "~2.2.1",
    "react-native-google-mobile-ads": "^6.3.0",
    "react-native-paper": "^4.12.1",
    "react-native-reanimated": "~2.8.0",
    "react-native-safe-area-context": "4.2.4",
    "react-native-screens": "~3.11.1",
    "react-native-svg": "12.3.0",
    "react-native-table-component": "^1.2.2",
    "react-native-web": "0.17.7",
    "sharp-cli": "^1.15.0"
  },
  "devDependencies": {
    "@babel/core": "^7.17.10",
    "@expo/webpack-config": "^0.16.24",
    "babel-plugin-transform-remove-console": "^6.9.4",
    "expo-cli": "^5.4.7"
  },
  "private": true
}

```

</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [x] **Android** but have not tested behavior on iOS
  - [ ] Both
- **Are you using `TypeScript`?**
  - `N`

</p>
</details>

---

- 👉 Check out [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

**Comment by mikehardy** [comment]
Hey there - using EAS build or managed workflow? managed can't load native modules. You have to use EAS build in order to add native code like this project. You don't specify

**Comment by shaun-ps-04** [comment]
Sorry I'm new to expo. My understanding is that EAS build is only to do with production builds not in development. How do I determine if it's a EAS build or managed workflow?

**Comment by daastrading** [comment]
I just init expo and using the latest expo and using bare workflow. Then I try to install this package and run "expo run:android". The project was build. But crash or stop when open the apps. Is there a problem with the package ? I see in the expo website that admob from expo is deprecated and recomend using this package. But not working the apps crash and stop when try to run.

**Comment by shaun-ps-04** [comment]
Is it possible to use this package in localhost?

**Comment by daastrading** [comment]
[](https://docs.page/invertase/react-native-google-mobile-ads)

I am missing this one :
**Attempting to build your app without a valid App ID may cause a crash during build.**

Now it is working

**Comment by shaun-ps-04** [comment]
> [Google AdMob](https://docs.page/invertase/react-native-google-mobile-ads)
> 
> I am missing this one : **Attempting to build your app without a valid App ID may cause a crash during build.**
> 
> Now it is working

Could you reproduce the steps please?

**Comment by mikehardy** [comment]
>  Is it possible to use this package in localhost?

I'm sorry, I don't know what that means.

I use it on my development machine, locally all the time though. it fetches ads (test ads) remotely though. is that localhost?

**Comment by shaun-ps-04** [comment]
Yes that's what I meant. 
I've created two other projects, 1 managed, 1 bare. But the results were the same. Perhaps I'm missing something.

Here are the steps I took:

```
expo init ad-module-test
expo install react-native-google-mobile-ads
```

Then I create the app on GoogleAdMob and retrieve the app id and insert them into the `app.json` like so:
```json
{
  "expo": {
    ...
  },
  "react-native-google-mobile-ads": {
    "android_app_id": "ca-app-pub-8695660704410804~8714442289",
    "ios_app_id": "ca-app-pub-8695660704410804~8714442289"
  }
}
```

And finally in `App.js`:

```javascript
...
import mobileAds from 'react-native-google-mobile-ads';

export default function App() {

  useEffect( () => {
    mobileAds()
      .initialize()
      .then( adapterStatuses => {} )
      .catch( err => {} )
  }, [])
  
  return (
    ...
  );
}

```

And yet the `initialize` method always raises this error:
> TypeError: null is not an object (evaluating 'RNAppModule.eventsNotifyReady')

Is there something I'm missing in this setup?

**Comment by mikehardy** [comment]
As mentioned elsewhere, I don't do expo personally. So in truth I have no idea. But your commands don't seem to line up with what I would expect based on this doc. Seems like you should be doing eas commands and building apks etc
https://docs.expo.dev/build-reference/apk/

**Comment by shaun-ps-04** [comment]
Thank you none the less for taking the time to help.
Last question then.
Do I need to build the app in order to test the ads?
Will it not work with `npm run start`?

**Comment by mikehardy** [comment]
It is my understanding that if you use expo, you need to use all expo commands and it's like being in a McDonalds where everything is McSomething, so you should be doing `expo ...` or `eas ...` and `npm ...` is something that makes me immediately suspicious. I'd carefully read the docs.

**Comment by mikehardy** [comment]
you might like: https://github.com/invertase/react-native-google-mobile-ads/issues/155#issuecomment-1142490214

**Comment by daastrading** [comment]
> > [Google AdMob](https://docs.page/invertase/react-native-google-mobile-ads)
> > I am missing this one : **Attempting to build your app without a valid App ID may cause a crash during build.**
> > Now it is working
> 
> Could you reproduce the steps please?

Expo have two type. 
1. Expo Manage Flow
2. Expo Bare Workflow.

For me. I recommend to use Bare Workflow since it can use Native Modules. Bare workflow also can use expo package.
**How to know if your expo is managed or bare ?**
_If you see android and ios folder like native react-native than it is a bare workflow._

**How to start with Expo Bare Workflow ?**
use :
`expo init test` command to create new project.
_than choose 'minimal' => this is a bare workflow._

**How to use '[react-native-google-mobile-ads](https://github.com/invertase/react-native-google-mobile-ads) ?**
This is the documentation in here : 
[Google Mobile Ads](https://docs.page/invertase/react-native-google-mobile-ads)

**What you should now ?**
use : 
# Install the admob module
yarn add react-native-google-mobile-ads
or npm i react-native-google-mobile-ads

Than to use the package is a must you should have Google Admob Account.
In the documentation : 
[Google Mobile Ads](https://docs.page/invertase/react-native-google-mobile-ads)

```
// <project-root>/app.json
{
  "react-native-google-mobile-ads": {
    "android_app_id": "ca-app-pub-xxxxxxxx~xxxxxxxx",
    "ios_app_id": "ca-app-pub-xxxxxxxx~xxxxxxxx"
  }
}
```

you must add the valid APP ID to **android_app_id** and **ios_app_id**.
If you do not specify the **Valid APP ID**. Than your project still success when build, but it will crash when you run it.
_So make sure to have an Admob Account before installing this package._

This is the example configuration in your expo **app.json**: 

```
{
  "expo": {
    "name": "testing",
    "slug": "testing",
    "description": "testing description",
    "version": "1.0.0",
    "privacy": "public",
    "notification": {
      "icon": "./assets/other/logo.png"
    },
    "icon": "./assets/other/logo.png",
    "splash": {
      "image": "./assets/other/splash.png",
      "resizeMode": "contain",
      "backgroundColor": "#034ea2"
    },    
    "platforms": [
      "ios",
      "android"
    ],
    "orientation": "portrait",
    "updates": {
      "fallbackToCacheTimeout": 0
    },
    "assetBundlePatterns": [
      "**/*"
    ],
    "ios": {
      "bundleIdentifier": "com.testing.testing",
      "buildNumber": "1.0.0"
    },
    "android": {
      "package": "com.testing.testing",
      "versionCode": 1
    }
  },
  "react-native-google-mobile-ads": {
    "android_app_id": "ca-app-pub-xxxxxx-xxxxx",
    "ios_app_id": "ca-app-pub-xxxxxx-xxxxx"
  }
}
```


Please note do not put **react-native-google-mobile-ads** inside "expo". See my example above.

Then how to use it in code ?
**The documentation is clear**. But make sure if you are in development mode. You can use : 

import { AppOpenAd, InterstitialAd, RewardedAd, BannerAd, TestIds } from 'react-native-google-mobile-ads';

```
# App Open
AppOpenAd.createForAdRequest(TestIds.APP_OPEN);

# Interstitial
InterstitialAd.createForAdRequest(TestIds.INTERSTITIAL);

# Rewarded
RewardedAd.createForAdRequest(TestIds.REWARDED);

# Banners
<BannerAd unitId={TestIds.BANNER} />
```

which **TestIds** is a testing for development only. After you are ready to publish to production change it to your **Ad unit ID** which can be create in your Google Admob Account in **Ad Units**.

**Comment by shaun-ps-04** [comment]
> you might like: [#155 (comment)](https://github.com/invertase/react-native-google-mobile-ads/issues/155#issuecomment-1142490214)

Thanks, unfortunately one of the commands was unavailable on windows so I couldn't get it to work.

However I managed to get it working by doing what [daastrading](https://github.com/daastrading) suggested above.
But so far I've only got it working on the simulator and not on my device because the linking for expo go isn't working.

```
expo init ad-test
expo install react-native-google-mobile-ads
``` 

Then configure `app.json`

```
expo run:android
```

I will do more testing on these things when I next have time, to see if I can get it to work on my device. Thank you both for your help.
